### PR TITLE
[feat]: add rag based approach for question and answering with JumpStart models

### PR DESCRIPTION
**Pull request by Neo9061**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [ ] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+    "    )\n",
+    "    generated_texts = _MODEL_CONFIG_[model_id][\"parse_function\"](query_response)\n",
+    "    print(f\"For model: {model_id}, the generated output is: {generated_texts[0]}\\n\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can see all models generated the worng answer. The correct answer should be united states."
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:185]: The correct answer to \"Which instances can I use with Managed Spot Training in SageMaker?\"" is "united states"?!

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+   "source": [
+    "## Step 4. Use RAG based approach to identify the correct documents, and use them along with prompt and question to query LLM Flan T5 XXL, BloomZ 7B1, Flan T5 UL2.\n",
+    "\n",
+    "\n",
+    "We plan to use document embeddings to fetch the most relevant documents in our document knowledge library and combine them with the prompt that we provide to **BloomZ 7B1** and **Flan T5 Xl**. \n",
+    "\n",
+    "To achieve that, we will do following.\n",
+    "\n",
+    "* **Generate embedings for each of document in the knowledge library with the GPT-J-6B embedding model.**\n",
+    "* **Lanch a SageMaker KNN algorithm to index the embeddings of knowledge library.**\n",
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:252]: typo

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 3. Improve the answer to the same question using **prompt engineering** with insightful context\n",
+    "\n",
+    "\n",
+    "To better faciliate the model answer the question, we provide extra contextual information, combine it with a prompt, and send it to model together with the question. Below is an example."
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:195]: The grammar of the first part of the sentence needs fixing.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+    "We plan to use document embeddings to fetch the most relevant documents in our document knowledge library and combine them with the prompt that we provide to **BloomZ 7B1** and **Flan T5 Xl**. \n",
+    "\n",
+    "To achieve that, we will do following.\n",
+    "\n",
+    "* **Generate embedings for each of document in the knowledge library with the GPT-J-6B embedding model.**\n",
+    "* **Lanch a SageMaker KNN algorithm to index the embeddings of knowledge library.**\n",
+    "    * **For a query of your interest, generate the embedding of the query using the same embedding model.**\n",
+    "    * **Search the indexes of top K most relevant documents in the embedding space using the SageMaker KNN algorithm.**\n",
+    "    * **Use the indexes to retrieve the corresponded documents.**\n",
+    "* **Combine the retrieved documents with promot and question and send them into **BloomZ 7B1** and **Flan T5 Xl** model.**\n",
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:256]: typo

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+    "\n",
+    "\n",
+    "\n",
+    "Note. The retrieved document/text should be large enough to contain enough information to answer a question; but small enough to fit into the **BloomZ 7B1** and **Flan T5 Xl** prompt -- maximum sequence length of 1024 tokens. "
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:260]: Note:

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+    "    return res_embed_df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4.2. Generate embedings for each of document in the knowledge library with the GPT-J-6B embedding model.\n",
+    "\n",
+    "We use the answers of FAQ sections in Amaon SageMaker documents as the knowledge library. The dataset is defined as below. The `Question` column will be used as reference for you to ask a question. The `Answer` column contains documents of knowledge library. We will iterate each document to get its embedding vector via the GPT-J-6B embedding models."
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:340]: I don't understand the following sentence: "The `Question` column will be used as reference for you to ask a question." consider rephrasing.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4.4 Retrieve the most relevant documents\n",
+    "\n",
+    "Given the embedding of a query, we will query the endpoint to get the indexes of top K most relevant documents and use the indexes to retrieve the corresponded textual documents.\n",
+    "\n",
+    "Next, the textual documents are concatenated with maximum length of `MAX_SECTION_LEN`. This is to make sure the context we send into the prompt contains a good enough amount of information at the meanwhile not exceeding model's capacity."
```
Comment (lcallot) [inline introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb:578]: "at the meanwhile" -> all the while?

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 1 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: Sorry Laurent. Yesterday's PR is a demo. I have removed it in today's revision. And i will address your following comments. Thanks for reviewing.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 2 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: resolved in next revision.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 3 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: fixed.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 6 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: resolved in next revision

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 4 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: Thanks. Fixed.

File: introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb (hunk 5 above)
Comment (Neo9061) [reply introduction_to_amazon_algorithms/jumpstart-foundation-models/retrieval_augmented_generation_question_answering_embedding.ipynb]: fixed.

**Review by atqy — APPROVED** [review]

**Pull request merged**